src/gax/src/error/core_error.rs

```diff
@@ -12,161 +12,664 @@
-    pub fn other<T: Into<BoxError>>(source: T) -> Self {
-        Error::new(ErrorKind::Other, source)
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// If true, the request was missing required parameters or the parameters
+    /// did not match any of the expected formats.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn is_binding(&self) -> bool {
+        matches!(&self.kind, ErrorKind::Binding(_))
     }
 
-    /// Returns the [ErrorKind] associated with this error.
-    pub fn kind(&self) -> ErrorKind {
-        self.kind.clone()
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// Cannot create the authentication headers.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn authentication(source: CredentialsError) -> Self {
+        Self {
+            kind: ErrorKind::Authentication(source),
+        }
     }
 
-    /// Recurses through the source error chain and returns a reference to the
-    /// inner value if it is of type `T`, or `None` if no such inner value is
-    /// found.
+    /// Not part of the public API, subject to change without notice.
     ///
-    /// # Examples
+    /// Could not create the authentication headers before sending the request.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn is_authentication(&self) -> bool {
+        matches!(self.kind, ErrorKind::Authentication(_))
+    }
+
+    /// Not part of the public API, subject to change without notice.
     ///
-    /// ```
-    /// # use google_cloud_gax::error::Error;
-    /// # use google_cloud_gax::error::HttpError;
-    /// use http::HeaderMap;
-    /// let error = Error::rpc(HttpError::new(404, HeaderMap::new(), None));
-    /// if let Some(e) = error.as_inner::<HttpError>() {
-    ///     assert_eq!(e.status_code(), 404);
-    /// }
-    /// ```
-    pub fn as_inner<T: std::error::Error + Send + Sync + 'static>(&self) -> Option<&T> {
-        let mut error = self.source.as_ref() as &(dyn std::error::Error);
-        loop {
-            match error.downcast_ref::<T>() {
-                Some(e) => return Some(e),
-                None => error = error.source()?,
+    /// A problem reported by the transport layer.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn http(status_code: u16, headers: HeaderMap, payload: bytes::Bytes) -> Self {
+        let kind = ErrorKind::Transport {
+            status_code: Some(status_code),
+            headers: Some(Box::new(headers)),
+            payload: Some(payload),
+            source: None,
+        };
+        Self { kind }
+    }
+
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// A problem in the transport layer without headers a full HTTP response.
+    ///
+    /// Examples include: a broken connection after the request is sent, or a
+    /// HTTP error that is *not*
```

Comment:
not ...?

The examples should match `is_io()`, no?

Reply:
Fixed.

I think we will need to make a second pass where we include that troubleshooting function in an example or as part of the user guide and expose enough predicates to make this error more useful.